tests/test_dimension.py

```diff
@@ -0,0 +1,61 @@
+import numpy as np
+
+from conftest import skipif_yask
+
+from devito import ConditionalDimension, Grid, TimeFunction, Eq, Operator, Constant
+
+
+@skipif_yask
+def test_conditional_basic():
+    nt = 19
+    grid = Grid(shape=(11, 11))
+    time = grid.time_dim
+
+    u = TimeFunction(name='u', grid=grid)
+    assert(grid.stepping_dim in u.indices)
+
+    u2 = TimeFunction(name='u2', grid=grid, save=nt)
+    assert(time in u2.indices)
+
+    factor = 4
+    time_subsampled = ConditionalDimension('t_sub', parent=time, factor=factor)
+    usave = TimeFunction(name='usave', grid=grid, save=(nt+factor-1)//factor,
+                         time_dim=time_subsampled)
+    assert(time_subsampled in usave.indices)
+
+    eqns = [Eq(u.forward, u + 1.), Eq(u2.forward, u2 + 1.), Eq(usave, u)]
+    op = Operator(eqns)
+    op.apply(t=nt)
+    assert np.all(u.data[(nt-1) % 3] == nt-1)
+    assert np.all(u2.data[i] == i for i in range(nt))
+    assert np.all(usave.data[i] == i*factor for i in range((nt+factor-1)//factor))
+
+
+@skipif_yask
+def test_conditional_shifted():
+    nt = 19
+    grid = Grid(shape=(11, 11))
+    time = grid.time_dim
+
+    u = TimeFunction(name='u', grid=grid)
+    assert(grid.stepping_dim in u.indices)
+
+    u2 = TimeFunction(name='u2', grid=grid, save=nt)
+    assert(time in u2.indices)
+
+    factor = 4
+    time_subsampled = ConditionalDimension('t_sub', parent=time, factor=factor)
+    usave = TimeFunction(name='usave', grid=grid, save=2, time_dim=time_subsampled)
+    assert(time_subsampled in usave.indices)
+
+    t_sub_shift = Constant(name='t_sub_shift', dtype=np.int32)
+
+    eqns = [Eq(u.forward, u + 1.), Eq(u2.forward, u2 + 1.),
+            Eq(usave.subs(time_subsampled, time_subsampled - t_sub_shift), u)]
+    op = Operator(eqns)
+
+    op.apply(time_s=10, time_e=nt, t_sub_shift=3)
```

Comment:
wait, doesn't this create negative indices?

first `time_subsampled` is `4/4 - 3 = -2`?

Reply:
right... @tjb900 ?

Reply:
Focussing, for a second, on why this wasn't caught during testing, this sort of a problem could be caught by the arguments "engine" right?

Reply:
There shouldn't be OOB access in this case because time_s=10, so the first matching index is 12, i.e. 12/4 - 3.

The subtraction is important, from my perspective, to be able to cross-correlate small ranges of time with each call - i.e. for checkpointing.

Reply:
right, dumb me!

Reply:
Oh ok didn't see that. I personaly would rather remove this one from the test base to avoid confusion, or at least put some comments above `op.apply()` like "starting at tim=10, time_sub - t_shift is in range".

Reply:
> I personaly would rather remove this one from the test base to avoid confusion

I disagree

But I'll leave a comment as you suggested (in the next PR at this point, which is imminent)